### PR TITLE
Feat/refactor matchmaker promise format

### PR DESCRIPTION
**Pull request by itschip**

### CODE REVIEW

File: phone/src/Phone.tsx [hunk 1]
```diff
@@ -63,11 +71,20 @@ function Phone() {
...
+  useEffect(() => {
+    if (!isEnvBrowser()) {
+      const resourceName = (window as any)?.GetParentResourceName() || 'npwd';
+      fetch(`https://cfx-nui-${resourceName}/config.json`).then(async (res) => {
+        const config = await res.json();
+        setResourceConfig(config);
+      });
+    }
+  }, []);
+
```
Comment (TasoOneAsia) [inline phone/src/Phone.tsx:78]: abstract this into a custom hook, lets not mount this in the main component

File: phone/src/apps/match/components/views/MatchList.tsx [hunk 2]
```diff
@@ -21,7 +22,8 @@ function MatchList() {
   const Nui = useNuiRequest();
   const classes = useStyles();
   const { t } = useTranslation();
-  const { matches, error } = useMatches();
+  const matches = useMatchesValue();
+  const error = useRecoilValue(matchState.errorLoadingMatches);
```
Comment (TasoOneAsia) [inline phone/src/apps/match/components/views/MatchList.tsx:26]: Might be better to just wrap the matchmaker app in an error boundary component and control error state handling that way

**Review by TasoOneAsia — COMMENTED** [review]
Quick review of a couple things until I get a chance to more fully review this

**Comment by RockySouthpaw** [comment]
Is this resolved in this PR as well? https://github.com/project-error/npwd/issues/272

File: resources/client/cl_match.ts [hunk 3]
```diff
@@ -8,44 +11,51 @@ import { sendMatchEvent } from '../utils/messages';
...
-RegisterNuiCallbackType(MatchEvents.GET_MATCHES);
-on(`__cfx_nui:${MatchEvents.GET_MATCHES}`, () => {
-  emitNet(MatchEvents.GET_MATCHES);
+RegisterNuiCB<Profile>(MatchEvents.UPDATE_MY_PROFILE, async (profile, cb) => {
+  const resp = await ClUtils.emitNetPromise<FormattedProfile>(
+    MatchEvents.UPDATE_MY_PROFILE,
+    profile,
+  );
+  cb({ resp });
 });
```
Comment (TasoOneAsia) [inline resources/client/cl_match.ts:43]: These can be proxies if you don't need to handle on client

**Comment by itschip** [comment]
> Is this resolved in this PR as well? #272

No, not yet, accidentally

File: phone/src/apps/match/components/MatchApp.tsx [hunk 4]
```diff
@@ -5,25 +5,33 @@ import { AppWrapper } from '../../../ui/components';
...
   const [activePage, setActivePage] = useState(0);
-  const { noProfileExists } = useProfile();
+  const [noProfileExists, setNoProfileExists] = useProfileExists();
+  const myProfile = useMyProfileValue();
 
   useEffect(() => {
-    Nui.send(MatchEvents.INITIALIZE);
-  }, [Nui]);
+    fetchNui(MatchEvents.INITIALIZE).then(() => console.log('Updated last activity'));
+  }, []);
```
Comment (TasoOneAsia) [inline phone/src/apps/match/components/MatchApp.tsx:26]: You said this was legacy? If so lets remove it

File: phone/src/apps/match/components/matches/Match.tsx [hunk 5]
```diff
@@ -22,6 +22,7 @@ interface IProps {
 }
 
 export const Match = ({ match }: IProps) => {
+  console.log('the match', match);
```
Comment (TasoOneAsia) [inline phone/src/apps/match/components/matches/Match.tsx:25]: Remove

File: phone/src/apps/match/components/views/MatchList.tsx (hunk 2 above)
Comment (TasoOneAsia) [reply phone/src/apps/match/components/views/MatchList.tsx:22]: Abstract with a hook within the state file to keep consistent with conventions

File: phone/src/apps/match/hooks/state.ts [hunk 6]
```diff
@@ -1,29 +1,98 @@
...
+import { FormattedMatch, FormattedProfile, MatchEvents } from '../../../../../typings/match';
+import { fetchNui } from '../../../utils/fetchNui';
+import { ServerPromiseResp } from '../../../../../typings/common';
+import LogDebugEvent from '../../../os/debug/LogDebugEvents';
+import { isEnvBrowser } from '../../../utils/misc';
+import { MockMatchesData, MockMyProfileData, MockProfilesData } from '../utils/constants';
 
 export const matchState = {
-  profiles: atom<FormattedProfile[]>({
+  profiles: atom<FormattedProfile[] | any>({
```
Comment (TasoOneAsia) [inline phone/src/apps/match/hooks/state.ts:10]: Why is this an any type?

File: phone/src/apps/match/hooks/useMatchActions.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  const setViewed = useCallback(
+    (id: number, liked: boolean) => {
+      setProfiles((profiles) =>
+        profiles.map((profile) => {
+          if (id === profile.id) return { ...profile, viewed: true, liked };
+          return profile;
+        }),
+      );
+
+      fetchNui<ServerPromiseResp<boolean>>(MatchEvents.SAVE_LIKES, [{ id, liked }]).then((resp) => {
```
Comment (TasoOneAsia) [inline phone/src/apps/match/hooks/useMatchActions.ts:33]: Is there any reason we are sending this in an array? That seems a little odd. If possible refactor to standard object.

File: phone/src/apps/match/hooks/useMatchActions.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  const { addAlert } = useSnackbar();
+
+  const formatProfile = (profile: Profile): FormattedProfile | null => {
+    return {
+      ...profile,
+      tagList: profile.tags.split(',').filter((t) => t), // remove any empty tags
+      lastActiveFormatted: dayjs.unix(profile.lastActive).format(t('DATE_TIME_FORMAT')),
+      viewed: false,
+    };
+  };
```
Comment (TasoOneAsia) [inline phone/src/apps/match/hooks/useMatchActions.ts:22]: This is a pure function that doesn't need state context, I would have it outside of the hook. Also is there any reason we have null as a possible return type?

File: phone/src/apps/match/hooks/useMatchActions.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        // Instead of just fetching the matches again, we might want to first format the profile to a 'Match',then to a 'FormattedMatch'
+        // and then just mutate the already existing state of matches?
+        // I guess that's for another time :P
+
+        if (resp.data) {
+          addAlert({
+            message: 'APPS_MATCH_NEW_LIKE_FOUND',
+            type: 'info',
+          });
+          fetchNui(MatchEvents.GET_MATCHES);
```
Comment (TasoOneAsia) [inline phone/src/apps/match/hooks/useMatchActions.ts:50]: Random `GetMatches` fetch without a handler for the promise response?

File: phone/src/os/apps/config/apps.tsx [hunk 10]
```diff
@@ -127,7 +128,13 @@ export const APPS: IAppConfig[] = [
...
     color: MATCH_APP_TEXT_COLOR,
     path: '/match',
-    Route: () => <AppRoute id="MATCH" path="/match" component={MatchApp} emitOnOpen={true} />,
+    Route: () => {
+      return (
+        <React.Suspense fallback={<LoadingSpinner />}>
+          <AppRoute id="MATCH" path="/match" component={MatchApp} emitOnOpen={true} />
+        </React.Suspense>
+      );
+    },
```
Comment (TasoOneAsia) [inline phone/src/os/apps/config/apps.tsx:137]: Is this necessary here? Try and move the suspense farther down in the tree. This here would render a transparent background while suspending.

**Review by TasoOneAsia — CHANGES_REQUESTED** [review]

File: phone/src/os/apps/config/apps.tsx (hunk 10 above)
Comment (itschip) [reply phone/src/os/apps/config/apps.tsx:137]: Need `MatchApp` wrapped in a new component in that case. Can do.

File: phone/src/apps/match/hooks/useMatchActions.ts (hunk 8 above)
Comment (itschip) [reply phone/src/apps/match/hooks/useMatchActions.ts:22]: This is probably not even used at all. Removing it.

File: phone/src/apps/match/hooks/useMatchActions.ts (hunk 7 above)
Comment (itschip) [reply phone/src/apps/match/hooks/useMatchActions.ts:33]: No clue at all. I wanted to ask Kire about it, but didn't get a response, so I left it be.

File: phone/src/os/apps/config/apps.tsx (hunk 10 above)
Comment (TasoOneAsia) [reply phone/src/os/apps/config/apps.tsx:137]: Or move the recoil value call in a component down the tree

**Pull request merged**